### PR TITLE
JAVA-3167 CompletableFutures.allSuccessful() may return never completed future

### PR DESCRIPTION
**Pull request by lukasz-antoniak**
Fix [JAVA-3167](https://datastax-oss.atlassian.net/jira/software/c/projects/JAVA/issues/JAVA-3167).

### CODE REVIEW

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/concurrent/CompletableFutures.java [hunk 1]
```diff
@@ -100,7 +100,10 @@ public static <T> CompletionStage<Void> allSuccessful(List<CompletionStage<T>> i
                 } else {
                   Throwable finalError = errors.get(0);
                   for (int i = 1; i < errors.size(); i++) {
-                    finalError.addSuppressed(errors.get(i));
```
Comment (tolbertam) [inline core/src/main/java/com/datastax/oss/driver/internal/core/util/concurrent/CompletableFutures.java:103]: Good catch, so it looks like without this, an IllegalArgumentException is thrown in the calling thread, right?:

```
java.lang.IllegalArgumentException: Self-suppression not permitted
	at java.base/java.lang.Throwable.addSuppressed(Throwable.java:1054)
	at com.datastax.oss.driver.internal.core.util.concurrent.CompletableFutures.lambda$allSuccessful$2(CompletableFutures.java:105)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2251)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:143)
	at com.datastax.oss.driver.internal.core.util.concurrent.CompletableFutures.allSuccessful(CompletableFutures.java:92)
	at com.datastax.oss.driver.internal.core.util.concurrent.CompletableFuturesTest.should_not_suppress_identical_exceptions(CompletableFuturesTest.java:22)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)
Caused by: java.lang.RuntimeException
	at com.datastax.oss.driver.internal.core.util.concurrent.CompletableFuturesTest.should_not_suppress_identical_exceptions(CompletableFuturesTest.java:14)
	... 27 more
```

**Review by tolbertam — APPROVED** [review]
👍 looks good

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/concurrent/CompletableFutures.java (hunk 1 above)
Comment (lukasz-antoniak) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/concurrent/CompletableFutures.java:103]: Exactly.

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/concurrent/CompletableFutures.java [hunk 2]
```diff
@@ -100,7 +100,10 @@ public static <T> CompletionStage<Void> allSuccessful(List<CompletionStage<T>> i
...
+                    Throwable suppressedError = errors.get(i);
+                    if (finalError != suppressedError) {
+                      finalError.addSuppressed(suppressedError);
+                    }
```
Comment (absurdfarce) [inline core/src/main/java/com/datastax/oss/driver/internal/core/util/concurrent/CompletableFutures.java:106]: Took me a minute to convince myself that != is correct over equals() here.  I couldn't find any docs on whether self-suppression was tied to reference equality or object equality but the more reference equality does seem like the more conservative approach here.

**Review by absurdfarce — APPROVED** [review]

**Pull request merged**